File: ios/SpentyAI/SpentyAI/Core/Models/CashFlowProjection.swift
import Foundation

struct CashFlowProjection: Codable {
    var monthlyRecurringIncome: Double?
    var monthlyRecurringExpense: Double?
    var monthlyMandateExpense: Double?
    var monthlyOdInterest: Double?
    var monthlyEmiTotal: Double?
    var monthlyNet: Double?
    var recurringItems: [RecurringItem]?
    var mandateItems: [MandateItem]?
    var odInterestItems: [ODInterestItem]?
    var emiItems: [EMIItem]?
    var projection: [ProjectionMonth]?
}

struct MandateItem: Codable, Identifiable {
    var id: String { mandateId ?? UUID().uuidString }
    var mandateId: String?
    var merchant: String?
    var amount: Double?
    var frequency: String?
    var mandateType: String?
    var monthlyEquivalent: Double?
    var source: String?
    var sourceEmailId: String?
    var sourceEmailSubject: String?
    var startDate: String?
    var debitDay: Int?

    /// Day of month for calendar placement: prefers explicit debitDay, falls back to startDate day
    var dayOfMonth: Int? {
        if let debitDay { return debitDay }
        guard let startDate else { return nil }
        let parts = startDate.split(separator: "-")
        guard parts.count >= 3, let day = Int(parts[2].prefix(2)) else { return nil }
        return day
    }
}

struct EMIItem: Codable, Identifiable {
    var id: String { accountId ?? UUID().uuidString }
    var accountId: String?
    var accountName: String?
    var emiAmount: Double?
    var emiDay: Int?
    var loanInterestRate: Double?
    var loanTenureMonths: Int?
}

struct ProjectionMonth: Codable, Identifiable {
    var id: String { label ?? "\(monthIndex ?? 0)" }
    var monthIndex: Int?
    var label: String?
    var projectedIncome: Double?
    var projectedExpense: Double?
    var mandateExpense: Double?
    var odInterest: Double?
    var net: Double?
    var runningBalance: Double?

    enum CodingKeys: String, CodingKey {
        case monthIndex = "month"
        case label
        case projectedIncome
        case projectedExpense
        case mandateExpense
        case odInterest
        case net
        case runningBalance
    }

    /// Convenience aliases so views can still use the old names
    var month: String? { label }
    var income: Double? { projectedIncome }
    var expense: Double? { projectedExpense }
    var mandates: Double? { mandateExpense }
    var cumulativeNet: Double? { runningBalance }
}

struct RecurringItem: Codable, Identifiable {
    var id: String { transactionId ?? description ?? UUID().uuidString }
    var transactionId: String?
    var description: String?
    var amount: Double?
    var frequency: String?
    var recurringFrequency: String?
    var recurrenceDate: Int?
    var transactionType: String?
    var monthlyAmount: Double?
    var categoryId: String?
    var accountId: String?
    var isRecurring: Bool?

    /// Convenience alias: the projection endpoint uses `frequency`, the list endpoint uses `recurring_frequency`
    var effectiveFrequency: String? { frequency ?? recurringFrequency }
    /// Convenience alias for type
    var type: String? { transactionType }
}

struct ODInterestItem: Codable, Identifiable {
    var id: String { accountId ?? UUID().uuidString }
    var accountId: String?
    var accountName: String?
    var monthlyInterest: Double?
    var balance: Double?
    var rate: Double?
    var interestChargeDay: Int?
}

File: android-native/app/src/main/java/com/spentyai/app/core/models/CashFlowProjection.kt
package com.spentyai.app.core.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class CashFlowProjection(
    val monthlyRecurringIncome: Double? = null,
    val monthlyRecurringExpense: Double? = null,
    val monthlyMandateExpense: Double? = null,
    val monthlyOdInterest: Double? = null,
    val monthlyEmiTotal: Double? = null,
    val monthlyNet: Double? = null,
    val recurringItems: List<RecurringItem>? = null,
    val mandateItems: List<MandateItem>? = null,
    val odInterestItems: List<ODInterestItem>? = null,
    val emiItems: List<EMIItem>? = null,
    val projection: List<ProjectionMonth>? = null
)

@Serializable
data class MandateItem(
    val mandateId: String? = null,
    val merchant: String? = null,
    val amount: Double? = null,
    val originalAmount: Double? = null,
    val currency: String? = null,
    val baseCurrency: String? = null,
    val exchangeRate: Double? = null,
    val isEstimatedRate: Boolean? = null,
    val frequency: String? = null,
    val mandateType: String? = null,
    val monthlyEquivalent: Double? = null,
    val source: String? = null,
    val sourceEmailId: String? = null,
    val sourceEmailSubject: String? = null,
    val startDate: String? = null,
    val debitDay: Int? = null
) {
    val id: String get() = mandateId ?: java.util.UUID.randomUUID().toString()

    val isForeignCurrency: Boolean
        get() {
            val c = currency ?: return false
            val b = baseCurrency ?: return false
            return !c.equals(b, ignoreCase = true)
        }

    val dayOfMonth: Int?
        get() {
            if (debitDay != null) return debitDay
            val sd = startDate ?: return null
            val parts = sd.split("-")
            if (parts.size >= 3) {
                return parts[2].take(2).toIntOrNull()
            }
            return null
        }
}

@Serializable
data class EMIItem(
    val accountId: String? = null,
    val accountName: String? = null,
    val emiAmount: Double? = null,
    val emiDay: Int? = null,
    val loanInterestRate: Double? = null,
    val loanTenureMonths: Int? = null
) {
    val id: String get() = accountId ?: java.util.UUID.randomUUID().toString()
}

@Serializable
data class ProjectionMonth(
    @SerialName("month") val monthIndex: Int? = null,
    val label: String? = null,
    val projectedIncome: Double? = null,
    val projectedExpense: Double? = null,
    val mandateExpense: Double? = null,
    val odInterest: Double? = null,
    val net: Double? = null,
    val runningBalance: Double? = null
) {
    val id: String get() = label ?: "${monthIndex ?: 0}"
    val month: String? get() = label
    val income: Double? get() = projectedIncome
    val expense: Double? get() = projectedExpense
    val mandates: Double? get() = mandateExpense
    val cumulativeNet: Double? get() = runningBalance
}

@Serializable
data class RecurringItem(
    val transactionId: String? = null,
    val description: String? = null,
    val amount: Double? = null,
    val frequency: String? = null,
    val recurringFrequency: String? = null,
    val recurrenceDate: Int? = null,
    val transactionType: String? = null,
    val monthlyAmount: Double? = null,
    val categoryId: String? = null,
    val accountId: String? = null,
    val isRecurring: Boolean? = null
) {
    val id: String get() = transactionId ?: description ?: java.util.UUID.randomUUID().toString()
    val effectiveFrequency: String? get() = frequency ?: recurringFrequency
    val type: String? get() = transactionType
}

@Serializable
data class ODInterestItem(
    val accountId: String? = null,
    val accountName: String? = null,
    val monthlyInterest: Double? = null,
    val balance: Double? = null,
    val rate: Double? = null,
    val interestChargeDay: Int? = null
) {
    val id: String get() = accountId ?: java.util.UUID.randomUUID().toString()
}

@Serializable
data class RecurringListResponse(
    val transactions: List<RecurringItem> = emptyList()
)

@Serializable
data class MandateListResponse(
    val mandates: List<Mandate> = emptyList()
)

@Serializable
data class UpcomingMandatesResponse(
    val upcoming: List<Mandate> = emptyList()
)

@Serializable
data class DetectMandatesResponse(
    val detected: List<DetectedMandatePattern>? = null,
    val total: Int? = null
)

@Serializable
data class DetectedMandatePattern(
    val merchant: String? = null,
    val amount: Double? = null,
    val frequency: String? = null,
    val occurrences: Int? = null,
    val lastDate: String? = null,
    val accountId: String? = null
)

@Serializable
data class MandateCreateBody(
    val merchant: String,
    val amount: Double,
    val frequency: String,
    val mandateType: String? = null
)

@Serializable
data class MandateUpdateBody(
    val merchant: String? = null,
    val amount: Double? = null,
    val status: String? = null,
    val frequency: String? = null,
    val mandateType: String? = null
)

@Serializable
data class ToggleRecurringBody(
    val isRecurring: Boolean,
    val recurringFrequency: String? = null
)
